ompi/mca/coll/tuned/coll_tuned_decision_fixed.c

```diff
@@ -48,48 +50,199 @@ ompi_coll_tuned_allreduce_intra_dec_fixed(const void *sbuf, void *rbuf, int coun
 {
-    size_t dsize, block_dsize;
-    int comm_size = ompi_comm_size(comm);
-    const size_t intermediate_message = 10000;
+    size_t dsize, total_dsize;
+    int communicator_size, alg;
+    communicator_size = ompi_comm_size(comm);
     OPAL_OUTPUT((ompi_coll_tuned_stream, "ompi_coll_tuned_allreduce_intra_dec_fixed"));
 
-    /**
-     * Decision function based on MX results from the Grig cluster at UTK.
+    ompi_datatype_type_size(dtype, &dsize);
+    total_dsize = dsize * (ptrdiff_t)count;
+
+    /** Algorithms:
+     *  {1, "basic_linear"},
+     *  {2, "nonoverlapping"},
+     *  {3, "recursive_doubling"},
+     *  {4, "ring"},
+     *  {5, "segmented_ring"},
+     *  {6, "rabenseifner"
      *
-     * Currently, linear, recursive doubling, and nonoverlapping algorithms
-     * can handle both commutative and non-commutative operations.
-     * Ring algorithm does not support non-commutative operations.
+     * Currently, ring, segmented ring, and rabenseifner do not support
+     * non-commutative operations.
      */
-    ompi_datatype_type_size(dtype, &dsize);
-    block_dsize = dsize * (ptrdiff_t)count;
-
-    if (block_dsize < intermediate_message) {
-        return (ompi_coll_base_allreduce_intra_recursivedoubling(sbuf, rbuf,
-                                                                 count, dtype,
-                                                                 op, comm, module));
-    }
-
-    if( ompi_op_is_commute(op) && (count > comm_size) ) {
-        const size_t segment_size = 1 << 20; /* 1 MB */
-        if (((size_t)comm_size * (size_t)segment_size >= block_dsize)) {
-            return (ompi_coll_base_allreduce_intra_ring(sbuf, rbuf, count, dtype,
-                                                        op, comm, module));
+    if( !ompi_op_is_commute(op) ) {
+        if (communicator_size < 4) {
+            if (total_dsize < 131072) {
+                alg = 3;
+            } else {
+                alg = 1;
+            }
+        } else if (communicator_size < 8) {
+            alg = 3;
+        } else if (communicator_size < 16) {
+            if (total_dsize < 1048576) {
+                alg = 3;
+            } else {
+                alg = 2;
+            }
+        } else if (communicator_size < 32) {
+            alg = 3;
+        } else if (communicator_size < 64) {
+            if (total_dsize < 131072) {
+                alg = 1;
+            } else {
+                alg = 3;
+            }
+        } else if (communicator_size < 128) {
+            if (total_dsize < 128) {
+                alg = 1;
+            } else if (total_dsize < 512) {
+                alg = 3;
+            } else if (total_dsize < 8192) {
+                alg = 1;
+            } else {
+                alg = 3;
+            }
+        } else if (communicator_size < 256) {
+            if (total_dsize < 2048) {
+                alg = 2;
+            } else if (total_dsize < 16384) {
+                alg = 1;
```

Comment:
I understand there is data to back this up, but I just can’t wrap my mind around the claim that An allreduce implemented as a reduce followed by a bcast Can have better performance than a recursive doubling implementation. We are talking about at least twice the latency, so even if the data is small, when you have 256 processes this should count for something.

Reply:
There's only one size where it outperforms:
From the data, it seems like recursive doubling doesn't do well until larger sizes
1 - linear
2 - non overlapping
3 - recursive doubling

I can't say I understand *why* this is the case, but I have the data for it.
```
Allreduce
#Nnodes    Message_size 0                    1                    2                    3                    4                    5                    6   
128        4            2913.64(4878.65)     1344.74(3104.49)     1259.40(3147.54)     3230.74(5481.80)     3002.00(4954.51)     2857.97(4782.51)     2826.67(4517.93)     2913.64(4878.65)    
128        8            2899.49(4864.18)     1316.23(3038.82)     1184.93(2940.13)     3150.57(5384.97)     3071.26(5012.43)     2879.02(4781.92)     2811.67(4574.71)     2899.49(4864.18)    
128        16           2866.31(4821.69)     1312.63(3046.77)     1182.64(2946.13)     3097.51(5284.69)     2958.97(4890.04)     2872.01(4829.44)     2810.05(4509.48)     2866.31(4821.69)    
128        32           2914.33(4893.33)     1308.55(3047.86)     1188.79(2947.79)     3155.50(5369.73)     2982.11(4939.97)     2580.43(4431.11)     2748.29(4489.69)     2914.33(4893.33)    
128        64           2905.83(4882.22)     1316.48(3050.56)     1184.62(2949.98)     3047.34(5201.51)     3024.41(5004.69)     2202.25(4277.14)     2673.65(4290.02)     2905.83(4882.22)    
128        128          2915.41(4892.76)     1326.50(3045.12)     1182.40(2937.46)     2886.84(4847.97)     2928.42(4899.58)     2143.79(4226.11)     1998.89(3815.83)     2915.41(4892.76)    
128        256          2713.00(4612.54)     1334.70(3033.14)     1188.84(2936.60)     2867.93(4816.76)     3015.24(5060.37)     2184.14(4284.24)     2181.36(4051.38)     2713.00(4612.54)    
128        512          3149.83(6395.99)     1380.31(3043.26)     1187.99(2936.51)     3485.17(5908.56)     13889.51(20102.06)   10352.55(16418.76)   4567.20(9209.34)     3149.83(6395.99)    
128        1024         3152.68(6385.89)     1418.47(3044.78)     1205.74(2950.83)     3540.21(5987.08)     14348.81(21068.35)   9486.46(15085.37)    4549.55(9041.11)     3152.68(6385.89)    
128        2048         3200.28(6455.68)     1476.17(3038.72)     3643.84(11269.21)    2918.98(5652.02)     15048.90(22636.64)   9004.92(18176.30)    5113.47(10475.57)    3200.28(6455.68)    
128        4096         3295.32(6555.58)     1679.44(3024.81)     6412.64(21712.51)    2848.15(5647.26)     13569.57(22881.78)   8980.50(18174.79)    5142.45(10647.91)    3295.32(6555.58)    
128        8192         3338.42(6519.30)     2100.74(3086.32)     2568.41(6543.05)     3011.61(5837.56)     13309.64(21974.66)   9287.97(18097.78)    5775.48(11949.52)    3338.42(6519.30)    
128        16384        11371.09(19503.20)   3747.43(3197.73)     1222.98(2309.60)     3065.43(5546.04)     13403.42(21779.87)   9568.73(17933.32)    5752.29(11851.30)    11371.09(19503.20)  
128        32768        9273.18(18127.16)    8865.84(11979.62)    1805.81(2428.45)     3486.90(5929.51)     10867.88(18496.86)   9302.34(18080.09)    6026.58(12574.90)    9273.18(18127.16)   
128        65536        9094.06(18570.70)    15196.44(16023.66)   4304.25(3354.22)     4532.85(7355.35)     8893.91(17602.65)    9359.78(18103.13)    6373.04(13157.22)    9094.06(18570.70)   
128        131072       9248.30(18648.35)    21599.01(19537.89)   7298.66(4853.51)     6158.44(8748.26)     9256.05(18023.50)    10528.02(18082.28)   6604.47(13622.15)    9248.30(18648.35)   
128        262144       10061.86(18937.07)   41759.74(44516.10)   11518.40(10318.94)   10147.57(14956.15)   12548.72(21517.08)   11036.23(18010.11)   6933.89(13948.64)    10061.86(18937.07)  
128        524288       154569.99(411382.35) 66071.82(46407.64)   13104.86(16139.88)   17745.35(22907.08)   73804.77(209112.42)  169329.06(434459.12) 8801.50(17421.10)    154569.99(411382.35)
128        1048576      77388.83(217694.92)  126609.06(95737.94)  24363.44(33186.15)   33791.94(40007.01)   30265.35(68950.84)   102546.42(366719.82) 12353.21(21203.30)   77388.83(217694.92) 
```

Reply:
Oh a small note, I am using similar logic to how the dynamic code works (For an unknown rank #, use the next smallest known rank #). The checks here select the previous smallest data value, so since we have data for 128 and 256, we use the 128 rank data for 128-255, 256 data for 256-511, etc.

Reply:
There is probably a better way to assign the data to ranks instead of ^this method.

Could do some midpoint logic...ie

If we have data for 32, 64, 128 nodes.

Data from 64 nodes would represent [(32 + 64)/2 , (128 + 64)/2) or [48, 96)
```
if (rank < 48) {
    use 32 rank data
} else if (rank < 96) {
    use 64 rank data
} else {
    use 128 rank data
}
```